File: src/input_manager.py
import csv
from pathlib import Path
from typing import List

class InputManager:
    """Parses CSV files and extracts a unique list of IDs."""

    def load_ids(self, file_path: Path) -> List[str]:
        """Reads the CSV and returns a list of unique strings from the first column."""
        if not file_path.exists():
            raise FileNotFoundError(f"Input file not found: {file_path}")

        ids = set()
        with open(file_path, mode='r', newline='', encoding='utf-8') as csvfile:
            # Detect header presence
            sample = csvfile.read(2048)
            csvfile.seek(0)
            has_header = False
            if sample:
                try:
                    has_header = csv.Sniffer().has_header(sample)
                except Exception:
                    has_header = False

            reader = csv.reader(csvfile)

            # Use a flag to ensure we only skip one header
            header_skipped = False
            if has_header:
                next(reader, None)
                header_skipped = True

            for row in reader:
                if row and row[0]:
                    val = row[0].strip()
                    if val:
                        # Fallback: if Sniffer failed but first row is "ID", skip it
                        if not header_skipped and val.upper() == "ID":
                            header_skipped = True
                            continue
                        ids.add(val)

        return sorted(list(ids))

    def validate_format(self, file_path: Path) -> bool:
        """Ensures the CSV exists and contains at least one column."""
        if not file_path.exists():
            return False

        try:
            with open(file_path, mode='r', newline='', encoding='utf-8') as csvfile:
                reader = csv.reader(csvfile)
                row = next(reader, None)
                return row is not None and len(row) >= 1
        except Exception:
            return False

File: src/main/java/com/oracle/tool/InputManager.java
package com.oracle.tool;

import java.io.IOException;
import java.io.Reader;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVParser;
import org.apache.commons.csv.CSVRecord;

/**
 * Parses CSV files and extracts a unique list of IDs.
 */
public class InputManager {

  /**
   * Reads the CSV and returns a list of unique strings from the first column.
   *
   * @param filePath Path to the CSV file.
   * @return Sorted list of unique IDs.
   * @throws IOException If an I/O error occurs or file not found.
   */
  public List<String> loadIds(Path filePath) throws IOException {
    if (!Files.exists(filePath)) {
      throw new IOException("Input file not found: " + filePath);
    }

    Set<String> ids = new HashSet<>();
    try (Reader reader = Files.newBufferedReader(filePath, StandardCharsets.UTF_8);
        CSVParser csvParser = new CSVParser(reader, CSVFormat.DEFAULT)) {

      boolean firstRecord = true;
      for (CSVRecord csvRecord : csvParser) {
        if (firstRecord) {
          firstRecord = false;
          continue;
        }
        if (csvRecord.size() > 0) {
          String val = csvRecord.get(0);
          if (val != null) {
            val = val.trim();
            if (!val.isEmpty()) {
              ids.add(val);
            }
          }
        }
      }
    }

    List<String> sortedIds = new ArrayList<>(ids);
    Collections.sort(sortedIds);
    return sortedIds;
  }

  /**
   * Ensures the CSV exists and contains at least one column.
   *
   * @param filePath Path to the CSV file.
   * @return True if valid, false otherwise.
   */
  public boolean validateFormat(Path filePath) {
    if (!Files.exists(filePath)) {
      return false;
    }

    try (Reader reader = Files.newBufferedReader(filePath, StandardCharsets.UTF_8);
        CSVParser csvParser = new CSVParser(reader, CSVFormat.DEFAULT)) {
      return csvParser.iterator().hasNext();
    } catch (Exception ex) {
      return false;
    }
  }
}
